### PR TITLE
HDDS-8560. TypedTable should store key/value codecs in its fields.

### PR DESCRIPTION
**Pull request by szetszwo**
## What changes were proposed in this pull request?

TypedTable looks up the same key/value codec from CodecRegistry for each serialization.  It should store the key/value codecs in its fields.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8560

## How was this patch tested?

By existing unit tests.

### CODE REVIEW

**Comment by guohao-rosicky** [comment]
Good optimization, so that Codec can be used repeatedly in TypedTable

**Comment by szetszwo** [comment]
@guohao-rosicky , thanks for taking a look!

**Comment by Galsza** [comment]
+1 The change looks good to me

**Review by adoroszlai — APPROVED** [review]
Thanks @szetszwo for the patch.  The change itself looks good.  I've noticed a minor issue introduced in HDDS-8552.  Please feel free to fix it in this change or the next one.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/TypedTable.java [hunk 1]
```diff
@@ -127,26 +144,20 @@ public void put(KEY key, VALUE value) throws IOException {
...
       // The buffers will be released after commit.
       rawTable.putWithBatch(batch,
           keyCodec.toDirectCodecBuffer(key),
           valueCodec.toDirectCodecBuffer(value));
     }
 
-    byte[] keyData = codecRegistry.asRawData(key);
-    byte[] valueData = codecRegistry.asRawData(value);
-    rawTable.putWithBatch(batch, keyData, valueData);
+    rawTable.putWithBatch(batch, encodeKey(key), encodeValue(value));
```
[the remark this answers is not in the document]
Comment (szetszwo) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/TypedTable.java:160]: Good catch!  Let me add `else`.

**Comment by adoroszlai** [comment]
Thanks @szetszwo for updating the patch.

**Pull request merged**